File: src/main/java/xyz/oribuin/lilori/managers/command/CommandEvent.java
package xyz.oribuin.lilori.managers.command;

import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.entities.*;
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent;
import net.dv8tion.jda.api.exceptions.ErrorResponseException;
import xyz.oribuin.lilori.LilOri;

import java.util.concurrent.TimeUnit;

public class CommandEvent {

    private final LilOri bot;
    private final GuildMessageReceivedEvent event;

    public CommandEvent(LilOri bot, GuildMessageReceivedEvent event) {
        this.bot = bot;
        this.event = event;
    }

    public GuildMessageReceivedEvent getEvent() {
        return event;
    }

    public void reply(String message) {
        event.getChannel().sendMessage(message).queue();
    }

    public void reply(EmbedBuilder embedBuilder) {
        event.getChannel().sendMessage(embedBuilder.build()).queue();
    }

    public void reply(MessageEmbed messageEmbed) {
        event.getChannel().sendMessage(messageEmbed).queue();
    }

    public void timedReply(String message, int time, TimeUnit timeUnit) {
        event.getChannel().sendMessage(message).queue(msg -> msg.delete().queueAfter(time, timeUnit));
    }

    public void timedReply(EmbedBuilder embedBuilder, int time, TimeUnit timeUnit) {
        event.getChannel().sendMessage(embedBuilder.build()).queue(msg -> msg.delete().queueAfter(time, timeUnit));
    }

    public void timedReply(MessageEmbed messageEmbed, int time, TimeUnit timeUnit) {
        event.getChannel().sendMessage(messageEmbed).queue(msg -> msg.delete().queueAfter(time, timeUnit));
    }

    public void deleteCmd() {
        event.getMessage().delete().queue();
    }

    public void deleteCmd(int time, TimeUnit timeUnit) {
        try {
            event.getMessage().delete().queueAfter(time, timeUnit);
        } catch (ErrorResponseException ignored) {
        }
    }

    public String getPrefix() {
        return bot.getGuildSettingsManager().getGuildSettings(event.getGuild()).getPrefix();
    }

    public Member getSelfMember() {
        return event.getGuild().getSelfMember();
    }

    public void kickMember(User user) {
        event.getGuild().getMember(user).kick().queue();
    }

    public void kickMember(User user, String reason) {
        event.getGuild().getMember(user).kick(reason).queue();
    }

    public void banMember(User user, int days) {
        event.getGuild().getMember(user).ban(days).queue();
    }

    public void banMember(User user, int days, String reason) {
        event.getGuild().getMember(user).ban(days, reason).queue();
    }

    public void banMember(Guild guild, User user, int days) {
        guild.getMember(user).ban(days).queue();
    }

    public void banMember(Guild guild, User user, int days, String reason) {
        guild.getMember(user).ban(days, reason).queue();
    }


    // GuildMessageReceivedEvent shortcuts

    public User getAuthor() {
        return event.getAuthor();
    }

    public MessageChannel getChannel() {
        return event.getChannel();
    }

    public TextChannel getTextChannel() {
        return (TextChannel) this.getChannel();
    }

    public Guild getGuild() {
        return event.getGuild();
    }

    public JDA getJDA() {
        return event.getJDA();
    }

    public Member getMember() {
        return event.getMember();
    }

    public Message getMessage() {
        return event.getMessage();
    }

    public PrivateChannel getPrivateChannel() {
        return event.getAuthor().openPrivateChannel().complete();
    }

    public long getResponseNumber() {
        return event.getResponseNumber();
    }
}

File: src/main/kotlin/xyz/oribuin/lilori/handler/CommandEvent.kt
package xyz.oribuin.lilori.handler

import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.entities.*
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent
import net.dv8tion.jda.api.exceptions.ErrorResponseException
import xyz.oribuin.lilori.LilOri
import xyz.oribuin.lilori.data.GuildSettings
import java.awt.Color
import java.net.URL
import java.util.concurrent.TimeUnit

class CommandEvent(private val bot: LilOri, val event: GuildMessageReceivedEvent) {

    fun reply(message: String) {
        event.channel.sendMessage(message).queue()
    }

    fun reply(embedBuilder: EmbedBuilder) {
        event.channel.sendMessage(embedBuilder.build()).queue()
    }

    fun reply(messageEmbed: MessageEmbed) {
        event.channel.sendMessage(messageEmbed).queue()
    }

    fun timedReply(message: String, time: Int, timeUnit: TimeUnit) {
        event.channel.sendMessage(message).queue { msg: Message -> msg.delete().queueAfter(time.toLong(), timeUnit) }
    }

    fun timedReply(embedBuilder: EmbedBuilder, time: Int, timeUnit: TimeUnit) {
        event.channel.sendMessage(embedBuilder.build()).queue { msg: Message -> msg.delete().queueAfter(time.toLong(), timeUnit) }
    }

    fun timedReply(messageEmbed: MessageEmbed, time: Int, timeUnit: TimeUnit) {
        event.channel.sendMessage(messageEmbed).queue { msg: Message -> msg.delete().queueAfter(time.toLong(), timeUnit) }
    }

    fun deleteCmd() {
        event.message.delete().queue()
    }

    fun deleteCmd(time: Int, timeUnit: TimeUnit) {
        try {
            event.message.delete().queueAfter(time.toLong(), timeUnit)
        } catch (ignored: ErrorResponseException) {
        }
    }

    fun getMemberFromString(member: String): Member? {
        return event.guild.getMemberByTag(member) ?: event.guild.getMembersByEffectiveName(member, true)[0] ?: event.guild.getMemberById(member)
    }

    val prefix: String
        get() = GuildSettings(guild).getPrefix()

    val color: Color
        get() = GuildSettings(guild).getColor()

    val selfMember: Member
        get() = event.guild.selfMember

    fun kickMember(user: User) {
        event.guild.getMember(user)?.kick()?.queue()
    }

    fun kickMember(user: User, reason: String) {
        event.guild.getMember(user)?.kick(reason)?.queue()
    }

    fun banMember(user: User, days: Int) {
        event.guild.getMember(user)?.ban(days)?.queue()
    }

    fun banMember(user: User, days: Int, reason: String) {
        event.guild.getMember(user)?.ban(days, reason)?.queue()
    }

    fun banMember(guild: Guild, user: User, days: Int) {
        guild.getMember(user)?.ban(days)?.queue()
    }

    fun banMember(guild: Guild, user: User, days: Int, reason: String) {
        guild.getMember(user)?.ban(days, reason)?.queue()
    }

    val args = event.message.contentRaw.split(" ").toTypedArray()

    // GuildMessageReceivedEvent shortcuts
    val author: User
        get() = event.author

    val channel: MessageChannel
        get() = event.channel

    val textChannel: TextChannel
        get() = channel as TextChannel

    val guild: Guild
        get() = event.guild

    val jda: JDA
        get() = event.jda

    val member: Member
        get() = event.member!!

    val message: Message
        get() = event.message

    val privateChannel: PrivateChannel
        get() = event.author.openPrivateChannel().complete()

    val responseNumber: Long
        get() = event.responseNumber

    fun sendEmbedReply(authorMessage: String, description: String) {
        val embedBuilder = EmbedBuilder()
                .setFooter("Created by Ori#0004", "http://img.oribuin.xyz/profile.png")
                .setColor(color)
                .setDescription(description)
                .setAuthor(authorMessage)

        event.channel.sendMessage(embedBuilder.build()).queue()
    }

    fun sendWebhook(url: URL, content: String) {
        // TODO
    }

    fun sendWebhook(url: URL, embed: MessageEmbed) {
        // TODO
    }

    fun sendWebhook(url: URL, content: String, embed: MessageEmbed) {
        // TODo
    }
}